Building zig fails on NetBSD-current amd64 and Aarch64

### Zig Version

0.12.0-dev.2517+3e939e615 (last working version)

### Steps to Reproduce and Observed Behavior

```
git pull
mkdir build; cd build
cmake -DCMAKE_BUILD_TYPE=Release -DZIG_NO_LIB=ON -GNinja ..
ninja install
```
Running on 
```
$ uname -a
NetBSD ymir.lorien.lan 10.99.10 NetBSD 10.99.10 (GENERIC) #27: Thu Feb  1 01:05:38 GMT 2024  sysbuild@ymir.lorien.lan:/dumps/sysbuild/amd64/obj/home/sysbuild/src/sys/arch/amd64/compile/GENERIC a
md64 x86_64 Intel 686-class NetBSD
```

with 
```
$ cmake --version
cmake version 3.28.2
```
and
```
$  ninja --version
1.11.1
```
as well as {llvm,clang,lld}-17.0.6.  

The result is as follows:
```

$ cmake -DCMAKE_BUILD_TYPE=Release -DZIG_NO_LIB=ON -GNinja ..
-- The C compiler identification is GNU 10.5.0
-- The CXX compiler identification is GNU 10.5.0
-- Detecting C compiler ABI info
$ cmake -DCMAKE_BUILD_TYPE=Release -DZIG_NO_LIB=ON -GNinja ..
-- The C compiler identification is GNU 10.5.0
-- The CXX compiler identification is GNU 10.5.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring zig version 0.12.0-dev.2753+3f5593205
-- Found llvm: -lLLVM-17;-lrt;-lexecinfo;-lpthread;-lm;-lz;-lzstd;-lterminfo;-lxml2 (Required is at least version "17")
-- Found clang: /usr/pkg/lib/libclang-cpp.so.17 (Required is at least version "17")
-- Found lld: /usr/pkg/lib/liblldMinGW.a;/usr/pkg/lib/liblldELF.a;/usr/pkg/lib/liblldCOFF.a;/usr/pkg/lib/liblldWasm.a;/usr/pkg/lib/liblldMachO.a;/usr/pkg/lib/liblldCommon.a (Required is at least
 version "17")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Configuring done (3.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ci4/src/zig/build

$🚩  ci4  …/zig/build () master $? ()  v10.5.0 () 10:39 ()
$ ninja install
[15/20] Running zig1.wasm to produce /home/ci4/src/zig/build/zig2.c
FAILED: zig2.c /home/ci4/src/zig/build/zig2.c
cd /home/ci4/src/zig && /home/ci4/src/zig/build/zig1 /home/ci4/src/zig/lib build-exe -ofmt=c -lc -OReleaseSmall --name zig2 -femit-bin="/home/ci4/src/zig/build/zig2.c" -target aarch64-netbsd --d
ep build_options --dep aro --mod root src/main.zig --mod build_options /home/ci4/src/zig/build/config.zig --mod aro_options src/stubs/aro_options.zig --mod Builtins/Builtin.def src/stubs/aro_bui
ltins.zig --mod Attribute/names.def src/stubs/aro_names.zig --mod Diagnostics/messages.def src/stubs/aro_messages.zig --dep build_options=aro_options --mod aro_backend deps/aro/backend.zig --dep
 Builtins/Builtin.def --dep Attribute/names.def --dep Diagnostics/messages.def --dep build_options=aro_options --dep backend=aro_backend --mod aro deps/aro/aro.zig
/lib/std/os.zig:4901:42: error: expected type 'u32', found 'c.O__struct_58429'
/lib/std/c.zig:398:23: note: struct declared here
/lib/std/c/netbsd.zig:18:47: note: parameter type declared here
/lib/std/c.zig:79:22: error: root struct of file 'c' has no member named 'padded_pthread_spin_t'
/lib/std/c.zig:1:1: note: struct declared here
ninja: build stopped: subcommand failed.
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring zig version 0.12.0-dev.2753+3f5593205
-- Found llvm: -lLLVM-17;-lrt;-lexecinfo;-lpthread;-lm;-lz;-lzstd;-lterminfo;-lxml2 (Required is at least version "17")
-- Found clang: /usr/pkg/lib/libclang-cpp.so.17 (Required is at least version "17")
-- Found lld: /usr/pkg/lib/liblldMinGW.a;/usr/pkg/lib/liblldELF.a;/usr/pkg/lib/liblldCOFF.a;/usr/pkg/lib/liblldWasm.a;/usr/pkg/lib/liblldMachO.a;/usr/pkg/lib/liblldCommon.a (Required is at least
 version "17")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Configuring done (3.7s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ci4/src/zig/build

$🚩  ci4  …/zig/build () master $? ()  v10.5.0 () 10:39 ()
$ ninja install
[15/20] Running zig1.wasm to produce /home/ci4/src/zig/build/zig2.c
FAILED: zig2.c /home/ci4/src/zig/build/zig2.c
cd /home/ci4/src/zig && /home/ci4/src/zig/build/zig1 /home/ci4/src/zig/lib build-exe -ofmt=c -lc -OReleaseSmall --name zig2 -femit-bin="/home/ci4/src/zig/build/zig2.c" -target aarch64-netbsd --d
ep build_options --dep aro --mod root src/main.zig --mod build_options /home/ci4/src/zig/build/config.zig --mod aro_options src/stubs/aro_options.zig --mod Builtins/Builtin.def src/stubs/aro_bui
ltins.zig --mod Attribute/names.def src/stubs/aro_names.zig --mod Diagnostics/messages.def src/stubs/aro_messages.zig --dep build_options=aro_options --mod aro_backend deps/aro/backend.zig --dep
 Builtins/Builtin.def --dep Attribute/names.def --dep Diagnostics/messages.def --dep build_options=aro_options --dep backend=aro_backend --mod aro deps/aro/aro.zig
/lib/std/os.zig:4901:42: error: expected type 'u32', found 'c.O__struct_58429'
/lib/std/c.zig:398:23: note: struct declared here
/lib/std/c/netbsd.zig:18:47: note: parameter type declared here
/lib/std/c.zig:79:22: error: root struct of file 'c' has no member named 'padded_pthread_spin_t'
/lib/std/c.zig:1:1: note: struct declared here
ninja: build stopped: subcommand failed.
```

### Expected Behavior

`ninja build `to complete. 

My last working versions under NetBSD-current are` 0.12.0-dev.2517+3e939e615 `under amd64 and` 0.12.0-dev.2507+abb8e7478` under Aarch64.

--- a/lib/std/c/dragonfly.zig
+++ b/lib/std/c/dragonfly.zig
@@ -12,7 +12,7 @@ pub fn _errno() *c_int {
 pub extern "c" fn getdents(fd: c_int, buf_ptr: [*]u8, nbytes: usize) c_int;
 pub extern "c" fn sigaltstack(ss: ?*stack_t, old_ss: ?*stack_t) c_int;
 pub extern "c" fn getrandom(buf_ptr: [*]u8, buf_len: usize, flags: c_uint) isize;
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 pub extern "c" fn arc4random_buf(buf: [*]u8, len: usize) void;
 
 pub const dl_iterate_phdr_callback = *const fn (info: *dl_phdr_info, size: usize, data: ?*anyopaque) callconv(.C) c_int;

--- a/lib/std/c/emscripten.zig
+++ b/lib/std/c/emscripten.zig
@@ -147,7 +147,7 @@ pub const pwritev64 = std.c.pwritev;
 pub const setrlimit64 = std.c.setrlimit;
 
 pub extern "c" fn sigaltstack(ss: ?*stack_t, old_ss: ?*stack_t) c_int;
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 pub extern "c" fn getentropy(buffer: [*]u8, size: usize) c_int;
 
 pub const pthread_attr_t = extern struct {

--- a/lib/std/c/freebsd.zig
+++ b/lib/std/c/freebsd.zig
@@ -15,7 +15,7 @@ pub extern "c" fn getrandom(buf_ptr: [*]u8, buf_len: usize, flags: c_uint) isize
 pub extern "c" fn pthread_getthreadid_np() c_int;
 pub extern "c" fn pthread_set_name_np(thread: std.c.pthread_t, name: [*:0]const u8) void;
 pub extern "c" fn pthread_get_name_np(thread: std.c.pthread_t, name: [*:0]u8, len: usize) void;
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 pub extern "c" fn arc4random_buf(buf: [*]u8, len: usize) void;
 
 pub extern "c" fn posix_memalign(memptr: *?*anyopaque, alignment: usize, size: usize) c_int;

--- a/lib/std/c/netbsd.zig
+++ b/lib/std/c/netbsd.zig
@@ -15,7 +15,7 @@ pub extern "c" fn dl_iterate_phdr(callback: dl_iterate_phdr_callback, data: ?*an
 
 pub extern "c" fn _lwp_self() lwpid_t;
 
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 pub extern "c" fn arc4random_buf(buf: [*]u8, len: usize) void;
 
 pub extern "c" fn __stat50(path: [*:0]const u8, buf: *Stat) c_int;
@@ -59,7 +59,7 @@ pub extern "c" fn posix_memalign(memptr: *?*anyopaque, alignment: usize, size: u
 pub extern "c" fn __msync13(addr: *align(std.mem.page_size) const anyopaque, len: usize, flags: c_int) c_int;
 pub const msync = __msync13;
 
-const pthread_spin_t = switch (builtin.cpu.arch) {
+pub const pthread_spin_t = switch (builtin.cpu.arch) {
     .aarch64, .aarch64_be, .aarch64_32 => u8,
     .mips, .mipsel, .mips64, .mips64el => u32,
     .powerpc, .powerpc64, .powerpc64le => i32,
@@ -70,7 +70,7 @@ const pthread_spin_t = switch (builtin.cpu.arch) {
     else => @compileError("undefined pthread_spin_t for this arch"),
 };
 
-const padded_pthread_spin_t = switch (builtin.cpu.arch) {
+pub const padded_pthread_spin_t = switch (builtin.cpu.arch) {
     .x86, .x86_64 => u32,
     .sparc, .sparcel, .sparc64 => u32,
     else => pthread_spin_t,

--- a/lib/std/c/openbsd.zig
+++ b/lib/std/c/openbsd.zig
@@ -14,7 +14,7 @@ pub extern "c" fn dl_iterate_phdr(callback: dl_iterate_phdr_callback, data: ?*an
 pub extern "c" fn arc4random_buf(buf: [*]u8, len: usize) void;
 
 pub extern "c" fn getthrid() pid_t;
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 
 pub extern "c" fn getdents(fd: c_int, buf_ptr: [*]u8, nbytes: usize) c_int;
 pub extern "c" fn sigaltstack(ss: ?*stack_t, old_ss: ?*stack_t) c_int;

--- a/lib/std/c/solaris.zig
+++ b/lib/std/c/solaris.zig
@@ -14,7 +14,7 @@ pub extern "c" fn dl_iterate_phdr(callback: dl_iterate_phdr_callback, data: ?*an
 
 pub extern "c" fn getdents(fd: c_int, buf_ptr: [*]u8, nbytes: usize) usize;
 pub extern "c" fn sigaltstack(ss: ?*stack_t, old_ss: ?*stack_t) c_int;
-pub extern "c" fn pipe2(fds: *[2]fd_t, flags: u32) c_int;
+pub extern "c" fn pipe2(fds: *[2]fd_t, flags: std.c.O) c_int;
 pub extern "c" fn arc4random_buf(buf: [*]u8, len: usize) void;
 pub extern "c" fn posix_memalign(memptr: *?*anyopaque, alignment: usize, size: usize) c_int;
 pub extern "c" fn sysconf(sc: c_int) i64;

--- a/lib/std/fs/Dir.zig
+++ b/lib/std/fs/Dir.zig
@@ -174,7 +174,7 @@ pub const Iterator = switch (builtin.os.tag) {
                     self.end_index = @as(usize, @intCast(rc));
                 }
                 const bsd_entry = @as(*align(1) posix.system.dirent, @ptrCast(&self.buf[self.index]));
-                const next_index = self.index + bsd_entry.reclen;
+                const next_index = self.index + if (@hasDecl(posix.system.dirent, "reclen")) bsd_entry.reclen() else bsd_entry.reclen;
                 self.index = next_index;
 
                 const name = @as([*]u8, @ptrCast(&bsd_entry.name))[0..bsd_entry.namlen];

